dashboard/app/models/levels/level.rb

```diff
@@ -376,6 +376,23 @@ def calculate_ideal_level_source_id
     update_attribute(:ideal_level_source_id, ideal_level_source.id) if ideal_level_source
   end
 
+  # Ensure Level uses script-cache for relevant queries.
+  def self.find(id)
+    find_by(id: id)
+  end
+
+  # Ensure Level uses script-cache for relevant queries.
+  def self.find_by(opts)
+    other = opts.dup
+    id = other.delete(:id)&.to_i
+    name = other.delete(:name)
+    if (id || name) && other.empty?
+      Script.cache_find_level(id || name)
+    else
+      super(opts)
+    end
+  end
```

Comment:
:cry: Overriding these ActiveRecord methods points out how often we're failing to think about perf.

Echoing your own comments here, if we're going in the direction of hiding caching logic behind an ActiveRecord-like interface, maybe we should go all-in and wrap the _whole_ query interface.  Maybe we can write a concern to do that.  [Maybe someone already has.](https://github.com/hooopo/second_level_cache)

I'll push this onto the list of tech debt for LP to consider in H1 2020.

Reply:
Good point @islemaster that this change as written sort of encourages not thinking about perf (except that you do have to think about it in order to call `.find_by` instead of `.find_by!`).

would you consider it an improvement to do a simple rename from all existing calls to `Level.find` to `Level.get_from_cache`, for example? and do a similar rename for `Level.find_by` --> `Level.find_by_from_cache`? That way, we would not have to remember as much in order to recognize new occurrences of ActiveRecord primitives as bypassing the cache.

@wjordan I would say please do not let this block your PR -- LP team can handle the cleanup.

Reply:
Yeah, I might - I've got this sense that I'd like us to override _all_ of the ActiveRecord query interface, or none of it, to make the behavior as unsurprising as possible.  Agreed on not blocking PR on this.